bandcamp-downloader.py

```diff
@@ -219,146 +226,262 @@ def main() -> int:
+        except Exception as e:
+            print_exception(e, 'An exception occurred trying to download album url [{}]:'.format(_url))
+    CONFIG['TQDM'].write("ERROR: Couldn't download pagedata for album at url [{}]".format(_url))
+    return {}
 
-def download_album(_album_url : str, _attempt : int = 1) -> str:
+# Download the file for the given bandcamp album, and notify TQDM when done.
+def download_and_log_album(_album : dict):
     try:
-        soup = BeautifulSoup(
-            requests.get(
-                _album_url,
-                cookies = CONFIG['COOKIE_JAR']
-            ).text,
-            'html.parser',
-            parse_only = SoupStrainer('div', id='pagedata'),
-        )
-        div = soup.find('div')
-        if not div:
-            CONFIG['TQDM'].write('ERROR: No div with pagedata found for album at url [{}]'.format(_album_url))
-            return
-
-        data = json.loads(html.unescape(div.get('data-blob')))
-        album = data['download_items'][0]['title']
-
-        if not 'downloads' in data['download_items'][0]:
-            CONFIG['TQDM'].write('WARN: Album [{}] at url [{}] has no downloads available.'.format(album, _album_url))
-            return
-
-        if not CONFIG['FORMAT'] in data['download_items'][0]['downloads']:
-            CONFIG['TQDM'].write('WARN: Album [{}] at url [{}] does not have a download for format [{}].'.format(album, _album_url, CONFIG['FORMAT']))
-            return
-
-        download_url = data['download_items'][0]['downloads'][CONFIG['FORMAT']]['url']
-        track_info = {key: data['download_items'][0][key] for key in TRACK_INFO_KEYS}
-        return download_file(download_url, track_info)
-    except IOError as e:
-        if _attempt < CONFIG['MAX_URL_ATTEMPTS']:
-            if CONFIG['VERBOSE'] >=2: CONFIG['TQDM'].write('WARN: I/O Error on attempt # [{}] to download the album at [{}]. Trying again...'.format(_attempt, _album_url))
-            time.sleep(CONFIG['URL_RETRY_WAIT'])
-            download_album(_album_url, _attempt + 1)
-        else:
-            print_exception(e, 'An exception occurred trying to download album url [{}]:'.format(_album_url))
+        download_album(_album)
     except Exception as e:
-        print_exception(e, 'An exception occurred trying to download album url [{}]:'.format(_album_url))
-    finally:
-        # only tell TQDM we're done on the first call
-        if _attempt == 1:
-            CONFIG['TQDM'].update()
-            time.sleep(CONFIG['POST_DOWNLOAD_WAIT'])
-
-def download_file(_url : str, _track_info : dict = None, _attempt : int = 1) -> str:
-    """Return a string representing the absolute path to the file being downloaded"""
+         band_name = _album.get('band_name', '')
+         title = _album.get('item_title', '')
+         print_exception(e, 'Trying to download album [{}] with artist [{}] and title [{}]:').format(
```

Comment:
You have a ctypo here. The `.format` call is on the `print_exception()` function instead of on the string.

Reply:
I sure do! Fixed ^,^